_posts/2025-11-06-sealing.markdown

```diff
@@ -0,0 +1,202 @@
+
+A lot of CHERIoT compartments use type-safe pointers as handles.
+For example, if you open a socket, you get back a sealed capability to that socket's state.
+The same thing happens for message queues, TLS sessions, and so on.
+
+In a conventional monolithic kernel, these would all be handles or file descriptors looked up in some table in the kernel.
+In a microkernel, they'd either be indexed per caller, or managed via some handle-manager service.
+In CHERIoT, these are simply opaque pointers.
+
+Exactly the same abstractions that you use for data hiding in good API design work between trust domains, with a little bit of hardware help.
+This is great for programmers because you can easily *retrofit* a security boundary.
+If you have built an API around exposing opaque types, turning it into a robust security boundary simply means doing a few checks on the public APIs.
+
+# Type safety gives static software-defined capabilities
+
+The same sealing mechanism that provides type safety for dynamic allocations also works for static objects.
+You can create a CHERIoT compartment that has access to a sealed object where another compartment owns the sealing key.
+The contents of these show up in the audit log.
+
+We use these throughout the system to implement *software* capabilities.
+CHERI (hardware) capabilities are unforgeable (delegable) tokens of authority that authorise access to memory.
+CHERIoT software capabilities are unforgeable (delegable) tokens of authority that authorise some other compartment to perform some action on your behalf.
+
+The first use of these most programmers will make is to allocate memory.
+If you call `malloc`, this is a thin compatibility wrapper around `heap_allocate`, which requires an allocator (software) capability as an argument.
+The allocator capability is a sealed object that contains a quota.
+It authorises you to allocate memory until your quota is exhausted.
+It also authorises you to free objects that were allocated from your quota and reclaim the quota.
+Similarly, when you create a connected TCP socket, you pass a capability that authorises you to connect to a specific host and port.
+
+These static capabilities can be inspected with [CHERIoT Audit](/rtos/firmware/auditing/2024/03/01/cheriot-audit.html) so you can write policies that say things like 'these four compartments, between them, can't allocate more than 16 KiB of RAM' or 'this compartment may connect to my cloud back end, but nowhere else'.
+
+# Sealed handles give trivial flow isolation
+
+Sealed capabilities are passed out and back as opaque values and can be unsealed without any global state beyond the sealing key.
+A lot of compartments that use them have no global mutable state.
+If you pass a sealed capability to a TLS session into the TLS compartment, that compartment can unseal it and then see the state of your TLS session.
+
+While operating on your TLS session, that compartment does not hold any capabilities to other TLS sessions.
```

Comment:
Mm, that's hand-wavier than I'd like but I don't know how to reword it succinctly.  A _thread_ operating in the compartment doesn't have access to other TLS sessions, but there might be multiple threads at once, and so "the compartment" in a sense has multiple handles opened at once.  Maybe something like...

```suggestion
The thread operating on _your_ TLS session, while within the TLS compartment, does not gain access to capabilities to other TLS sessions.
```